### PR TITLE
[KYUUBI #1008] Implement Hadoop File System Delegation Token Provider

### PR DESCRIPTION
**Pull request by zhouyifan279**
### _Why are the changes needed?_
This PR finishes the work of task #1008 

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -985,6 +985,12 @@
                 <version>${hadoop.version}</version>
             </dependency>
 
+            <dependency>
+                <groupId>org.apache.hadoop</groupId>
+                <artifactId>hadoop-hdfs</artifactId>
```
Comment (pan3793) [inline pom.xml:990]: We don't need it, it's already covered by `hadoop-client-api` and `hadoop-client-runtime`

Comment (zhouyifan279) [reply pom.xml:990]: Checked that hadoop-hdfs is included in org.apache.hadoop:hadoop-client-minicluster.
Removed hadoop-hdfs dependency here.

**Comment by zhouyifan279** [comment]
@pan3793 @yaooqinn Ready to be reviewed.

File: kyuubi-server/pom.xml [hunk 2]
```diff
@@ -164,6 +164,12 @@
             <groupId>org.apache.hbase</groupId>
             <artifactId>hbase-client</artifactId>
             <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.apache.hadoop</groupId>
+                    <artifactId>hadoop-common</artifactId>
+                </exclusion>
+            </exclusions>
```
Comment (pan3793) [inline kyuubi-server/pom.xml:172]: Waiting #1028, would be ready today.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+class HadoopFsDelegationTokenProvider extends HadoopDelegationTokenProvider with Logging {
+
+  override val serviceName: String = "hadoopfs"
+
+  private var internalConf: Configuration = _
+
+  override def delegationTokensRequired(
+      hadoopConf: Configuration,
+      kyuubiConf: KyuubiConf): Boolean = {
+    UserGroupInformation.isSecurityEnabled
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala:40]: shall we call `UserGroupInformation.setConfiguration` before ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  override def obtainDelegationTokens(
+      hadoopConf: Configuration,
+      kyuubiConf: KyuubiConf,
+      owner: String,
+      creds: Credentials): Unit = {
+    // A FileSystem object are cached in FileSystem.CACHE by a composite key.
+    // The UserGroupInformation object used to create it is part of that key.
+    // If cache is enabled, new FileSystem objects are cached at every method invocation.
+    if (internalConf == null) {
+      internalConf = disableFsCache(kyuubiConf, hadoopConf)
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala:52]: why we cache `internalConf ` but not cache `filesystem`. isn't it should be same ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala (hunk 3 above)
Comment (zhouyifan279) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala:40]: Change `UserGroupInformation#conf` might be a little risky. How about `SecurityUtil.getAuthenticationMethod(hadoopConf) != AuthenticationMethod.SIMPLE`?

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala:40]: sgtm

File: kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala (hunk 4 above)
Comment (zhouyifan279) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala:52]: Number of cached FileSystem objects grows as kyuubi user number grows. To reduce memory usage under large number of users, FileSystem objects are closed after obtaining tokens.

Besides, although cached FileSystem objects maitains connections to NameNodes, connections are closed in 10 seconds by default. Token renewal won't benefit from cached FileSystem objects.

**Comment by pan3793** [comment]
#1028 got merged, it excluded all hadoop deps from hudi, pls revert the related changes and rebase on latest master branch

**Comment by zhouyifan279** [comment]
@pan3793 Thanks for the notification. Rebased on the master branch.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 5]
```diff
@@ -249,6 +249,22 @@ object KyuubiConf {
...
       .createWithDefault(Duration.ofMinutes(1).toMillis)
 
+  val CREDENTIALS_HADOOP_FS_ENABLED: ConfigEntry[Boolean] =
+    buildConf("credentials.hadoopfs.enabled")
+      .doc("Whether to renew HadoopFS DelegationToken")
+      .version("1.4.0")
+      .booleanConf
+      .createWithDefault(true)
+
+  val CREDENTIALS_HADOOP_FS_URLS: ConfigEntry[Seq[String]] = KyuubiConf
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:259]: rm KyuubiConf

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 6]
```diff
@@ -249,6 +249,22 @@ object KyuubiConf {
...
+    .buildConf("credentials.hadoopfs.urls")
+    .doc("Extra Hadoop filesystem URLs for which to request delegation tokens. " +
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:261]: indent

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 7]
```diff
@@ -249,6 +249,22 @@ object KyuubiConf {
...
 
+  val CREDENTIALS_HADOOP_FS_ENABLED: ConfigEntry[Boolean] =
+    buildConf("credentials.hadoopfs.enabled")
+      .doc("Whether to renew HadoopFS DelegationToken")
+      .version("1.4.0")
+      .booleanConf
+      .createWithDefault(true)
+
+  val CREDENTIALS_HADOOP_FS_URLS: ConfigEntry[Seq[String]] =
+    buildConf("credentials.hadoopfs.urls")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:260]: How about
```
credentials.hadoop.filesystems
credentials.hadoop.enabled

```

**Review by yaooqinn — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 7 above)
Comment (zhouyifan279) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:260]: In reference to the [discussion](https://github.com/apache/spark/pull/16432#issuecomment-270320823) about the serviceName of `HadoopFSCredentialProvider` in Spark, using `hadoopfs` to stand for Hadoop Compatible file systems seems to be more appropriate.

Maybe we, should rename `credentials.hadoopfs.urls` to `credentials.hadoopfs.uris` as they are parsed as java.net.URI objects.

**Review by yaooqinn — APPROVED** [review]
LGTM

File: kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    val internalConf = disableFsCache(kyuubiConf, hadoopConf)
+
+    doAsProxyUser(owner) {
+      val fileSystems =
+        HadoopFsDelegationTokenProvider.hadoopFSsToAccess(kyuubiConf, internalConf)
+
+      try {
+        fileSystems.foreach { fs =>
+          info(s"getting token owned by $owner for: $fs")
+          fs.addDelegationTokens(null, creds)
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala:62]: why not use kyuubi server user `UserGroupInformation.getCurrentUser().getUserName()` as the renewer ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+  def doAsProxyUser[T](proxyUser: String)(f: => T): T = {
+    // For some reason the Scala-generated anonymous class ends up causing an
+    // UndeclaredThrowableException, even if you annotate the method with @throws.
+    try {
+      val ugi = UserGroupInformation.createProxyUser(proxyUser, UserGroupInformation.getCurrentUser)
+      ugi.doAs(new PrivilegedExceptionAction[T] {
+        override def run(): T = f
+      })
+    } catch {
+      case e: UndeclaredThrowableException => throw Option(e.getCause).getOrElse(e)
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala:105]: what happen if we fail to get token for one user, does it affects other user ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala (hunk 8 above)
Comment (zhouyifan279) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala:62]: We actually obtain a new token every time. Renewer is not needed.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala (hunk 9 above)
Comment (zhouyifan279) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala:105]: `HadoopCredentialsManager` will catch the exception and schedule a retry task. 

Failures are controlled in user scope and will not affect other users.

**Comment by zhouyifan279** [comment]
@ulysses-you modified according to advice. Ready to be reviewed again.

**Review by ulysses-you — APPROVED** [review]
LGTM, pending GA

**Comment by ulysses-you** [comment]
thanks, merging to master

**Pull request closed without merging**